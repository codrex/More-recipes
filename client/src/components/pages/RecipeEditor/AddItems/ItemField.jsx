import React from 'react';
import { Field } from 'redux-form';
import PropTypes from 'prop-types';

const ItemField = ({
  Component,
  name,
  placeholder,
  clearValidationError,
  validate,
  validationError
}) => (
  <Field
    component={Component}
    name={name}
    type="text"
    id={name}
    placeholder={placeholder}
    fgClassName="d-flex flex-column no-margin"
    className="no-margin add-item-input"
    validate={validate}
    onFocus={() => clearValidationError({
      [`${name}s`]: ''
    })}
    externalError={{
      touched: true,
      error: validationError
    }}
  />
);

ItemField.propTypes = {
  validate: PropTypes.func.isRequired,
  clearValidationError: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  Component: PropTypes.oneOfType(
    [
      PropTypes.node,
      PropTypes.element,
      PropTypes.func
    ]
  ).isRequired,
  validationError: PropTypes.oneOfType(
    [
      PropTypes.string,
      PropTypes.array, PropTypes.bool
    ]
  ).isRequired
};

export default ItemField;
