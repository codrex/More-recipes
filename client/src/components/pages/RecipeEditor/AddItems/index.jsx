import React from 'react';
import PropTypes from 'prop-types';
import ItemsList from './ItemsList';
import { Input, Textarea, Form } from '../../../common/FormElements';
import { validateItem } from '../../../../utils/validator';
import ItemField from './ItemField';
import { RECIPE_ADDED } from '../../../../constants';

/**
 * React component to add recipe Items
 */
class AddItems extends React.Component {
  /**
   * constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      items: props.items || [],
      showInput: true,
      validationError: props.externalError
    };
  }

  /**
   * @param {Object} nextProps
   * @return {boolean} boolean value
   */
  componentWillReceiveProps = (nextProps) => {
    if (this.state.validationError !== nextProps.externalError) {
      this.setState({
        validationError: nextProps.externalError
      });
    }
    this.setState({
      items: nextProps.items
    });
  }

  /**
   * @param {object} nextProps
   * @return {boolean} boolean value
   */
  shouldComponentUpdate = (nextProps) => {
    if (nextProps.message === RECIPE_ADDED) {
      return false;
    }
    return true;
  }

  /**
   * add item to list
   * @param {String} value
   * @return {undefined} undefined
   */
  addItem = (value) => {
    const { directions, ingredients } = this.props;
    const { items } = this.state;
    const item = value[this.props.name].toLowerCase();

    const itemsUpdate = (ingredients && [item].concat(items)) ||
    (directions && items.concat(item));

    this.props.sendItemsToStore(itemsUpdate);
    this.props.initialize();
  }

  /**
   * remove item from list
   * @param {number} index
   * @return {undefined} undefined
   */
  deleteFromList = (index) => {
    const items = [].concat(this.state.items);
    items.splice(index, 1);
    this.props.sendItemsToStore(items);
  }

  /**
   * edit item
   * @param {string} value
   * @param {number} index
   * @return {undefined} undefined
   */
  editItems = (value, index) => {
    const items = [].concat(this.state.items);
    items[index] = value;
    this.props.sendItemsToStore(items);
  }

  /**
   * validate item
   * @return {Array} array of strings
   * @param {string} value
   */
  validate = (value) => {
    const { name } = this.props;
    const lowerCaseValue = value && value.toLowerCase();
    if (this.state.items.includes(lowerCaseValue)) {
      return [`${name} is already on the list`];
    }
    return validateItem(lowerCaseValue, name);
  }

  /**
   * render form
   * @return {React} react component
   */
  renderForm = () => {
    const {
      handleSubmit,
      name,
      placeholder,
      ingredients,
      directions,
      clearValidationError
    } = this.props;
    const Component = (ingredients && Input) || (directions && Textarea);
    if (this.state.showInput) {
      return (
        <Form
          onSubmit={handleSubmit(this.addItem)}
          submitBtnText={`add ${name}`}
          secondary
          lg={false}
          id={`${name}-submit-btn`}
        >
          <ItemField
            Component={Component}
            name={name}
            placeholder={placeholder}
            validate={this.validate}
            clearValidationError={clearValidationError}
            validationError={this.state.validationError}
          />
        </Form>
      );
    } return null;
  }

  /**
   * render list
   * @return {React} react component
   */
  renderList = () => {
    const {
      name,
      Component,
      ingredients,
      directions
    } = this.props;
    return (
      <ItemsList
        items={this.state.items}
        deleteItem={this.deleteFromList}
        editItem={this.editItems}
        componentType={name}
        Component={Component}
        name={name}
        directions={directions}
        ingredients={ingredients}
        uniquenessChecker={this.validate}
      />
    );
  }

  /**
   * render
   * @return {React} react component
   */
  render = () => {
    const { name } = this.props;
    const { items } = this.state;
    return (
      <div className="col-12 items ">
        <div className="items-header">
          <h4 className="items-header-text">
            {`${name}s ${' '}${items.length}`}
          </h4>
          {this.renderForm()}
        </div>
        <div className="col-12 items-list-wrapper">
          {this.renderList()}
        </div>
      </div>
    );
  }
}

AddItems.defaultProps = {
  ingredients: false,
  directions: false,
  externalError: [],
  Component: null
};

AddItems.propTypes = {
  clearValidationError: PropTypes.func.isRequired,
  directions: PropTypes.bool,
  externalError: PropTypes.oneOfType(
    [
      PropTypes.string,
      PropTypes.array, PropTypes.bool
    ]
  ),
  handleSubmit: PropTypes.func.isRequired,
  ingredients: PropTypes.bool,
  initialize: PropTypes.func.isRequired,
  items: PropTypes.arrayOf(PropTypes.shape).isRequired,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  sendItemsToStore: PropTypes.func.isRequired,
  Component: PropTypes.oneOfType(
    [
      PropTypes.node,
      PropTypes.element,
      PropTypes.func
    ]
  )
};

export default AddItems;
