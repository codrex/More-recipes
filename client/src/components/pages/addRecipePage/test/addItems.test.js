import React from 'react';
import AddItems from '../addItems/addItems';
import renderer from 'react-test-renderer';

const props = {
  handleSubmit: jest.fn(),
  placeholder: '',
  name: 'name',
  initialize: jest.fn(),
};

describe('Add items component ', () => {
  test('render as expected when directions is passed as props', () => {
    const component = renderer.create(
      <AddItems {...props} directions />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('render as expected when ingredients is pass as props', () => {
    const component = renderer.create(
      <AddItems {...props} ingredients />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
