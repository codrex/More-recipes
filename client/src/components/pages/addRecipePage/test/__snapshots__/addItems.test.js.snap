// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add items component  render as expected when directions is passed as props 1`] = `
<div
  className="col-xs-12 col-sm-12 col-11 items "
>
  <div
    className="items-header"
  >
    <h4
      className="lead items-header-text"
    >
      names  0
      <div
        className="icon-wrapper"
        onClick={[Function]}
        onMouseEnter={undefined}
      >
        <i
          className="icon fa fa-plus float-right"
          data-target={undefined}
          data-toggle={undefined}
        >
          <span />
        </i>
      </div>
    </h4>
  </div>
  <div
    className="col-12 items-list-wrapper"
  >
    <div />
  </div>
</div>
`;

exports[`Add items component  render as expected when ingredients is pass as props 1`] = `
<div
  className="col-xs-12 col-sm-12 col-11 items "
>
  <div
    className="items-header"
  >
    <h4
      className="lead items-header-text"
    >
      names  0
      <div
        className="icon-wrapper"
        onClick={[Function]}
        onMouseEnter={undefined}
      >
        <i
          className="icon fa fa-plus float-right"
          data-target={undefined}
          data-toggle={undefined}
        >
          <span />
        </i>
      </div>
    </h4>
  </div>
  <div
    className="col-12 items-list-wrapper"
  >
    <div />
  </div>
</div>
`;
